File: YouboraLib/comm/YBRequest.m
//
//  YBRequest.m
//  YouboraLib
//
//  Created by Joan on 16/03/2017.
//  Copyright © 2017 NPAW. All rights reserved.
//

#import "YBRequest.h"
#import "YBLog.h"
#import "YBConstants.h"

// Required for User-Agent building
#include <sys/sysctl.h>
#import <UIKit/UIKit.h>

NSString * const YouboraHTTPMethodGet = @"GET";
NSString * const YouboraHTTPMethodPost = @"POST";
NSString * const YouboraHTTPMethodHead = @"HEAD";
NSString * const YouboraHTTPMethodOptions = @"OPTIONS";
NSString * const YouboraHTTPMethodPut = @"PUT";
NSString * const YouboraHTTPMethodDelete = @"DELETE";
NSString * const YouboraHTTPMethodTrace = @"TRACE";

@interface YBRequest()

/// ---------------------------------
/// @name Private properties
/// ---------------------------------

@property(nonatomic, strong) NSMutableArray<YBRequestSuccessBlock> * successListenerList;
@property(nonatomic, strong) NSMutableArray<YBRequestErrorBlock> * errorListenerList;
@property(nonatomic, assign) unsigned int pendingAttempts;

@end

@implementation YBRequest

static NSMutableArray<YBRequestSuccessBlock> * everySuccessListenerList;
static NSMutableArray<YBRequestErrorBlock> * everyErrorListenerList;

#pragma mark - Init
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.successListenerList = [NSMutableArray arrayWithCapacity:1];
        self.errorListenerList = [NSMutableArray arrayWithCapacity:1];
        self.listenerParams = [[NSDictionary alloc] init];
        
        self.maxRetries = 3;
        self.retryInterval = 5000;
        self.method = YouboraHTTPMethodGet;
        self.body = @"";
    }
    return self;
}

- (instancetype) initWithHost:(nullable NSString *) host andService:(nullable NSString *) service {
    self = [self init];
    self.host = host;
    self.service = service;
    return self;
}

#pragma mark - Public methods
-(void)send {
    self.pendingAttempts = self.maxRetries + 1;
    [self sendRequest];
}

-(NSURL *)getUrl {
    //NSURLComponents * components = [NSURLComponents new];
    NSURLComponents * components = [NSURLComponents componentsWithString:self.host];
    //components.host = self.host;
    if (self.service) {
        components.path = self.service;
    }
    
    // Build query params
    if (self.params != nil && self.params.count > 0) {
        
        NSMutableArray<NSURLQueryItem *> * queryItems = [NSMutableArray arrayWithCapacity:self.params.count];
        
        for (NSString * key in self.params) {
            NSString * value = self.params[key];
            if (value != nil) {
                // Avoid sending null values
                NSURLQueryItem * queryItem = [NSURLQueryItem queryItemWithName:key value:value];
                [queryItems addObject:queryItem];
            }
        }
        
        [components setQueryItems:queryItems];
        components.percentEncodedQuery = [[components.percentEncodedQuery stringByReplacingOccurrencesOfString:@"+" withString:@"%2B"] stringByReplacingOccurrencesOfString:@"%20" withString:@"+"];
    }
    
    return components.URL;
}

- (void)setParam:(NSString *)value forKey:(NSString *)key {
    if (value) {
        if (self.params == nil) {
            self.params = [NSMutableDictionary dictionaryWithObject:value forKey:key];
        } else {
            
            [self.params setObject:value forKey:key];
        }
    }
}

- (NSString *)getParam:(NSString *)key {
    return [self.params objectForKey:key];
}

#pragma mark - Private methods
- (NSMutableURLRequest *) createRequestWithUrl:(NSURL *) url {
    return [NSMutableURLRequest requestWithURL:url];
}

- (void) sendRequest {
    self.pendingAttempts--;
    @try {
        // Create request object
        NSMutableURLRequest * request = [self createRequestWithUrl:[self getUrl]];
        
        if ([YBLog isAtLeastLevel:YBLogLevelVerbose]) {
            [YBLog requestLog:@"XHR Req: %@", request.URL.absoluteString];
            if(self.body != nil && ![self.body isEqualToString:@""]){
                if([self.method isEqualToString:YouboraHTTPMethodPost]){
                    [YBLog debug:[NSString stringWithFormat:@"Req body: %@", self.body]];
                }
            }
        }
        
        // Set request headers if any
        if (self.requestHeaders != nil && self.requestHeaders.count > 0) {
            [request setAllHTTPHeaderFields:self.requestHeaders];
        }
        
        request.HTTPMethod = self.method;
        
        if(self.body != nil && ![self.body isEqualToString:@""] && [self.method isEqualToString:YouboraHTTPMethodPost]){
            [request setHTTPBody:[self.body dataUsingEncoding:NSUTF8StringEncoding]];
        }
        
        // User-agent
        [request setValue:[self getUserAgent] forHTTPHeaderField:@"User-Agent"];
        
        // Send request
        __weak typeof(self) weakSelf = self;
        NSURLSession * session = [NSURLSession sharedSession];
        [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
            __strong typeof(weakSelf) strongSelf = self;
            if (strongSelf == nil) {
                [YBLog error:@"YBRequest instance has been deallocated while waiting for completion handler"];
                return;
            }
            
            if(response != nil) {
                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                [YBLog debug:[NSString stringWithFormat:@"Response code for: %@ %ld",weakSelf.service, (long)httpResponse.statusCode]];
            }
            
            if (error == nil) {
                [weakSelf didSucceedWithData:data andResponse:response];
            } else {
                [weakSelf didFailWithError:error];
            }
        }] resume];
        
    } @catch (NSException *exception) {
        [YBLog logException:exception];
        [self didFailWithError:nil];
    }
}

- (void) didSucceedWithData:(NSData *) data andResponse:(NSURLResponse *) response {
    for (YBRequestSuccessBlock block in everySuccessListenerList) {
        @try {
            block(data, response, self.listenerParams);
        } @catch (NSException *exception) {
            [YBLog logException:exception];
        }
    }
    
    for (YBRequestSuccessBlock block in self.successListenerList) {
        @try {
            block(data, response, self.listenerParams);
        } @catch (NSException *exception) {
            [YBLog logException:exception];
        }
    }
}

- (void) didFailWithError:(NSError *) error {
    
    // Callbacks
    for (YBRequestErrorBlock block in everyErrorListenerList) {
        @try {
            block(error);
        } @catch (NSException *exception) {
            [YBLog logException:exception];
        }
    }
    
    for (YBRequestErrorBlock block in self.errorListenerList) {
        @try {
            block(error);
        } @catch (NSException *exception) {
            [YBLog logException:exception];
        }
    }
    
    // Retry
    if (self.pendingAttempts > 0) {
        [YBLog warn:[NSString stringWithFormat:@"Request \"%@\" failed. Retry %d of %d in %dms.", self.service, (self.maxRetries + 1 - self.pendingAttempts), self.maxRetries, self.retryInterval]];
        //__weak typeof (self) weakSelf = self;
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(self.retryInterval * NSEC_PER_MSEC)), dispatch_get_main_queue(), ^{
            //__strong typeof (weakSelf) strongSelf = weakSelf;
            [self sendRequest];
        });
    } else {
        [YBLog error:[NSString stringWithFormat:@"Aborting failed request \"%@\". Max retries reached(%d)", self.service, self.maxRetries]];
    }
}

/**
 * Builds a string with the User-Agent header content.
 *
 * This method will also normalize the user agent string. With "normalize"
 * we mean to convert the user-agent string to latin chars with no diacritic
 * signs Normalization could be necessary when the app bundle has non-latin
 * chars. It's been observed that, with Japanese characters, the User-Agent
 * header was arriving empty to the backend
 * @returns The User-Agent built string.
 */
- (NSString *) getUserAgent {
    
    static NSString * builtUserAgentNormalized;
    
    // Build User-Agent only once
    static dispatch_once_t onceTokenUA;
    dispatch_once(&onceTokenUA, ^{
        size_t size;
        
        // Set 'oldp' parameter to NULL to get the size of the data
        // returned so we can allocate appropriate amount of space
        sysctlbyname("hw.machine", NULL, &size, NULL, 0);
        
        char* name = (char*)malloc(size);
        
        // Get the platform name
        sysctlbyname("hw.machine", name, &size, NULL, 0);
        
        // Place name into a string
        NSString *machine = [NSString stringWithCString:name encoding:NSUTF8StringEncoding];
        
        free(name);
        
        UIDevice * device = [UIDevice currentDevice];
        
        NSMutableString * builtUserAgent = [NSMutableString stringWithFormat:@"%@/%@/%@/%@/%@",
                                            [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleName"],
                                            [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"],
                                            [device model],
                                            machine,
                                            [device systemVersion]
                                            ];
        
        // Transform non-latin to latin and remove diacritical signs
        CFMutableStringRef inputRef = (__bridge CFMutableStringRef) builtUserAgent;
        CFStringTransform(inputRef, NULL, kCFStringTransformToLatin, false); // transform to latin chars
        CFStringTransform(inputRef, NULL, kCFStringTransformStripCombiningMarks, false); // get rid of diacritical signs
        
        builtUserAgentNormalized = builtUserAgent;
    });
    
    return builtUserAgentNormalized;
}

#pragma mark - Static methods
- (void) addRequestSuccessListener:(YBRequestSuccessBlock) successBlock{
    if (successBlock != nil && ![self.successListenerList containsObject:successBlock]) {
        [self.successListenerList addObject:successBlock];
    }
}
- (void) removeRequestSuccessListener:(YBRequestSuccessBlock) successBlock{
    if (successBlock != nil) {
        [self.successListenerList removeObject:successBlock];
    }
}
- (void) addRequestErrorListener:(YBRequestErrorBlock) errorBlock{
    if (errorBlock != nil && ![self.errorListenerList containsObject:errorBlock]) {
        [self.errorListenerList addObject:errorBlock];
    }
}
- (void) removeRequestErrorListener:(YBRequestErrorBlock) errorBlock{
    if (errorBlock != nil) {
        [self.errorListenerList removeObject:errorBlock];
    }
}
+ (void) addEveryRequestSuccessListener:(YBRequestSuccessBlock) successBlock{
    if (successBlock != nil) {
        if (everySuccessListenerList == nil) {
            everySuccessListenerList = [NSMutableArray arrayWithObject:successBlock];
        } else {
            [everySuccessListenerList addObject:successBlock];
        }
    }
}
+ (void) removeEveryRequestSuccessListener:(YBRequestSuccessBlock) successBlock{
    if (everySuccessListenerList != nil && successBlock != nil) {
        [everySuccessListenerList removeObject:successBlock];
    }
}
+ (void) addEveryRequestErrorListener:(YBRequestErrorBlock) errorBlock{
    if (errorBlock != nil) {
        if (everyErrorListenerList == nil) {
            everyErrorListenerList = [NSMutableArray arrayWithObject:errorBlock];
        } else {
            [everyErrorListenerList addObject:errorBlock];
        }
    }
}
+ (void) removeEveryRequestErrorListener:(YBRequestErrorBlock) errorBlock{
    if (everyErrorListenerList != nil && errorBlock != nil) {
        [everyErrorListenerList removeObject:errorBlock];
    }
}
@end

File: YouboraLib/YBRequest.swift
//
//  YBRequest.swift
//  YouboraLib
//
//  Created by Tiago Pereira on 03/07/2020.
//  Copyright © 2020 NPAW. All rights reserved.
//

import Foundation
#if os(iOS)
import UIKit
#endif

/**
* Type of the success closure
*
*  - data: (Data *) the data as returned by the completionHandler.
*  - response: (URLResponse *) the response as returned by the completionHandler.
*/
@objcMembers public class YBRequestSuccess: NSObject {
    let closure: (_ data: Data, _ response: URLResponse, _ listenerParams: NSMutableDictionary) -> Void
    
    init(closure: @escaping (_ data: Data, _ response: URLResponse, _ listenerParams: NSMutableDictionary) -> Void) {
        self.closure = closure
    }
}

/**
* Type of the error block
*
*  - error: (Error *) error as returned by the completionHandler.
*/
@objcMembers public class YBRequestError: NSObject {
    let closure: (_ error: Error) -> Void
    
    init(closure: @escaping (_ error: Error) -> Void) {
        self.closure = closure
    }
}

/**
* Class with all the http methods used by Youbora
*/
@objcMembers class YouboraHTTPMethod: NSObject {
    static let get = "GET"
    static let post = "POST"
    static let head = "HEAD"
    static let put = "PUT"
    static let delete = "DELETE"
    static let trace = "TRACE"
}

@objcMembers class YBRequest: NSObject {
    /// The host where the YBRequest is performed to
    var host: String

    /// The service. This will be the "/something" part of the url.
    /// For instance the "/start" in "a-fds.youborafds01.com/start"
    var service: String?

    /// NSDictionary with params to add to the http request
    var params: NSMutableDictionary?

    /// NSDictionary with Request Headers to add to the http request
    var requestHeaders: [ String: String]?

    /// The retry interval for this request. In milliseconds. The default value is 5000.
    let retryInterval: UInt

    /// The number of retries for this request. The default value is 3.
    var maxRetries: UInt

    /// Method of the HTTP request. Default is <YouboraHTTPMethodGet>
    var method: String

    /// In case of wanting some params back. Default empty
    var listenerParams: NSMutableDictionary

    /// Request body in case of being method POST
    var body: String
    
    /// Closure to indicate success request
    fileprivate var successListenerList: [YBRequestSuccess?]
    
    /// Closure to indicate error on request
    fileprivate var errorListenerList: [YBRequestError?]
    
    fileprivate static var everySuccessListenerList: [YBRequestSuccess]?
    fileprivate static var everyErrorListenerList: [YBRequestError]?
    
    lazy var userAgent:String = {
        
        // Get machine name
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let machine = machineMirror.children.reduce("") { identifier, element in
          guard let value = element.value as? Int8, value != 0 else { return identifier }
          return identifier + String(UnicodeScalar(UInt8(value)))
        }

        var builtUserAgent: String?
        
        #if os(iOS)
        let device = UIDevice.current

        if let object = Bundle.main.infoDictionary?["CFBundleName"],
            let object1 = Bundle.main.infoDictionary?["CFBundleShortVersionString"] {
            builtUserAgent = "\(object)/\(object1)/\(device.model)/\(machine)/\(device.systemVersion)"
        }
        #else

        var size = 0
        sysctlbyname("hw.model", nil, &size, nil, 0)
        var modelName = [CChar](repeating: 0,  count: size)
        sysctlbyname("hw.model", &modelName, &size, nil, 0)

        let machineName = String(cString: modelName)

        if let object = Bundle.main.infoDictionary?["CFBundleName"],
            let object1 = Bundle.main.infoDictionary?["CFBundleShortVersionString"] {
            builtUserAgent = "\(object)/\(object1)/Mac/\(machineName)/\(ProcessInfo.processInfo.operatingSystemVersion)"
        }


        #endif
        
        guard let tmpBuiltUserAgent = builtUserAgent else {
            return ""
        }
        
        let mutableString = NSMutableString(string: tmpBuiltUserAgent) as CFMutableString
        
        CFStringTransform(mutableString, nil, kCFStringTransformToLatin, false) // transform to latin chars
        CFStringTransform(mutableString, nil, kCFStringTransformStripCombiningMarks, false) // get rid of diacritical signs

        return tmpBuiltUserAgent
    }()
    
    /// indicates the number of times that YBRequest tried to do a request
    var pendingAttemps: UInt

    init(host: String, service: String?) {
        self.host = host
        self.service = service
        self.maxRetries = 3
        self.retryInterval = 5000
        self.body = ""
        self.listenerParams = [:]
        self.method = YouboraHTTPMethod.get
        self.pendingAttemps = 0
        self.successListenerList = Array<YBRequestSuccess?>(repeating: nil, count: 1)
        self.errorListenerList = Array<YBRequestError?>(repeating: nil, count: 1)
    }
    
    /**
    * Sends this Request over the network.
    */
    public func send() {
        self.pendingAttemps = self.maxRetries+1
        self.sendRequest()
    }
    
    /**
    * Builds the url. It consists of the following: <host> + <service> + query params
    * @returns the full query url
    */
    
    public func getUrl() -> URL? {
        var components = URLComponents(string: self.host)
        
        if let service = self.service {
            components?.path = service
        }
        
        if let params = self.params {
            components?.queryItems = params.compactMap { (key: Any, value: Any) -> URLQueryItem? in
                guard let key = key as? String, let value = value as? String else {
                    return nil
                }
                return URLQueryItem(name: key, value: value)
            }
            
            let percentageEncodedQuery = components?.percentEncodedQuery?.replacingOccurrences(of: "+", with: "%2B").replacingOccurrences(of: "%20", with: "+")
            
            components?.percentEncodedQuery = percentageEncodedQuery
            
        }
        
        return components?.url
    }
    
    /**
     * Sets one key-value pair onto the params
     * @param key param key
     * @param value param value
     */
    public func setParam(_ value: String, _ key: String) {
        if self.params == nil {
            self.params = [:]
        }
        
        self.params?[key] = value
    }
    
    /**
     * Returns the param value for the given key
     * @param key the key for the desired value
     * @returns the param value
     */
    
    public func getParam(key: String) -> String? {
        return params?[key] as? String
    }
    
    private func sendRequest() {
        guard let url = self.getUrl() else {
                return
        }
        
        self.pendingAttemps -= 1
        
        var request = URLRequest(url: url)
        if YBSwiftLog.isAtLeastLevel(.verbose) {
            YBSwiftLog.requestLog("XHR Req: %@", url.absoluteString)
            if self.body != "" && self.method == YouboraHTTPMethod.post {
                YBSwiftLog.debug("Req body: %@", self.body)
            }
        }
        
        // Set request headers if any
        request.allHTTPHeaderFields = self.requestHeaders
        
        request.httpMethod = self.method
        
        if self.body != "" && self.method == YouboraHTTPMethod.post {
            request.httpBody = self.body.data(using: .utf8)
        }
        
        request.addValue(self.getUserAgent(), forHTTPHeaderField: "User-Agent")
        
        // Send request
        URLSession.shared.dataTask(with: request) {[weak self] (data, response, error) in
            guard let strongSelf = self else {
                YBSwiftLog.error("YBRequest instance has been deallocated while waiting for completion handler")
                return
            }
            
            if let response = response as? HTTPURLResponse,
                let service = strongSelf.service{
                YBSwiftLog.debug("Response code for: %@ %ld", service, response.statusCode)
            }
            
            if let error = error {
                self?.didFail(error: error)
            } else {
                self?.didSucceed(data: data, response: response)
            }
        }.resume()
    }
    
    /**
    * Builds a string with the User-Agent header content.
    *
    * This method will also normalize the user agent string. With "normalize"
    * we mean to convert the user-agent string to latin chars with no diacritic
    * signs Normalization could be necessary when the app bundle has non-latin
    * chars. It's been observed that, with Japanese characters, the User-Agent
    * header was arriving empty to the backend
    * @returns The User-Agent built string.
    */
    
    private func getUserAgent() -> String {
        return self.userAgent
    }
    
    private func didSucceed(data: Data?, response: URLResponse?) {
        guard let data = data,
            let response = response else {
                return
        }
        
        for successListener in self.successListenerList {
            successListener?.closure(data, response, self.listenerParams)
        }
    }
    
    private func didFail(error: Error) {
        for errorListener in self.errorListenerList {
            errorListener?.closure(error)
        }
        
        if self.pendingAttemps > 0 {
            if let service = self.service {
                YBSwiftLog.warn("Request \"%@\" failed. Retry %d of %d in %dms.", service, (self.maxRetries + 1 - self.pendingAttemps), self.maxRetries, self.retryInterval)
            }
            
            let deadlineTime = DispatchTime.now() + .seconds(Int(self.retryInterval))
            
            DispatchQueue.main.asyncAfter(deadline: deadlineTime) {
                self.sendRequest()
            }
        } else {
            if let service = self.service {
                YBSwiftLog.error("Aborting failed request \"%@\". Max retries reached(%d)", service, self.maxRetries)
            }
        }
    }
    
    /**
     * Adds a success listener.
     * @param successBlock the listener to add.
     */
    public func addRequestSuccessListener(_ success: YBRequestSuccess) {
        if !self.successListenerList.contains(success) {
            self.successListenerList.append(success)
        }
    }
    
    /**
     * Removes a success listener
     * @param successBlock the listener ot remove
     */
    public func removeRequestSuccessListener(_ success: YBRequestSuccess) {
        if let index = self.successListenerList.index(of: success) {
            self.successListenerList.remove(at: index)
        }
    }
    
    /**
     * Adds an error listener. These listeners will be called for each failed retry. If you want to
     * implement you own retry management, do so here and <setMaxRetries:> to 0.
     * @param errorBlock the listener to add
     */
    public func addRequestErrorListener(_ error: YBRequestError) {
        if !self.errorListenerList.contains(error) {
            self.errorListenerList.append(error)
        }
    }
    /**
     * Remove an error listener
     * @param errorBlock the listener to remove
     */
    public func removeRequestErrorListener(_ error: YBRequestError) {
        if let index = self.errorListenerList.index(of: error) {
            self.errorListenerList.remove(at: index)
        }
    }
    
    /// ---------------------------------
    /// @name Static methods
    /// ---------------------------------
    /**
     * Adds a <b>global</b> success listener. These listeners will be called <b>for all the Requests</b>.
     * @param successBlock the global listener to add.
     */
    static public func addEveryRequestSuccessListener(_ success: YBRequestSuccess) {
        if var everySuccessListenerList = everySuccessListenerList {
            everySuccessListenerList.append(success)
        } else {
            everySuccessListenerList = Array(arrayLiteral: success)
        }
    }
    
    /**
     * Removes a global success listener.
     * @param successBlock the global listener to remove
     */
    static public func removeEveryRequestSuccessListener(_ success: YBRequestSuccess) {
        if let index = everySuccessListenerList?.index(of: success) {
            everySuccessListenerList?.remove(at: index)
        }
    }
    
    /**
     * Adds a <b>global</b> error listener. These listeners will be called <b>for all the Requests</b>.
     * @param errorBlock the global listener to add.
     */
    static public func addEveryRequestErrorListener(_ error: YBRequestError) {
        if var everyErrorListenerList = everyErrorListenerList {
            everyErrorListenerList.append(error)
        } else {
            everyErrorListenerList = Array(arrayLiteral: error)
        }
    }
    
    /**
     * Removes a global error listener.
     * @param errorBlock the global listener to remove
     */
    static public func removeEveryRequestErrorListener(_ error: YBRequestError) {
        if let index = everyErrorListenerList?.index(of: error) {
            everyErrorListenerList?.remove(at: index)
        }
    }
}
